#**Sentiment analysis using Simple RNN**

> This notebook is part of my learning journey in deep learning fundamentals.
> Focus is on understanding concepts rather than achieving maximum accuracy.

# Objective

This notebook aims to demonstrate the process of building and evaluating a **Recurrent Neural Network (RNN)** using **<font color=cyan>TensorFlow</font>** to classify movie reviews as positive or negative (**Sentiment Analysis**). Specifically, I have used the IMDB dataset from **<font color=cyan>keras</font>** library.
The major layers included in this RNN model are **Embedding layer**, **SimpleRNN layer** and **Dense layer**.

#**Importing Libraries**

In [ ]:
import tensorflow as tf
from tensorflow import keras
from keras import layers, models

import warnings
warnings.filterwarnings('ignore')

#**<font color='red'>1.</font> Loading Dataset**

In [ ]:
vocab_size = 10000

(X_train, y_train), (X_test, y_test) = keras.datasets.imdb.load_data(num_words=vocab_size)

print(f"X_train length : {len(X_train)}")
print(f"X_test length : {len(X_test)}")
print("\nX_train sample : ")
print(X_train[0][:20])
print("\nLabel:", y_train[0])

X_train length : 25000
X_test length : 25000

X_train sample : 
[1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25]

Label: 1


#**<font color='red'>2.</font> Pre-processing**


##**Padding**
Each review is a sequence of integers representing words. These sequences have different lengths.

Neural networks require inputs of equal length. We pad all sequences to a fixed length using zeros. This ensures consistent input shape for training.

In [ ]:
max_length = 200

X_train_padded = keras.preprocessing.sequence.pad_sequences(X_train, maxlen=max_length)
X_test_padded = keras.preprocessing.sequence.pad_sequences(X_test, maxlen=max_length)

print(X_train_padded.shape)

(25000, 200)


#**<font color='red'>3.</font> Train the RNN Model**

In [ ]:
# Define the basic RNN Model
rnn = models.Sequential([
    layers.Embedding(input_dim=vocab_size, output_dim=32, input_length=max_length),
    layers.SimpleRNN(32),
    layers.Dense(1, activation='sigmoid')
])

# Compile the model
rnn.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)

In [ ]:
# Model training
history = rnn.fit(
    X_train_padded, y_train,
    epochs=3,
    batch_size=64,
    validation_split=0.2
)

Epoch 1/3
313/313 ━━━━━━━━━━━━━━━━━━━━ 20s 55ms/step - accuracy: 0.6579 - loss: 0.6211 - val_accuracy: 0.6902 - val_loss: 0.5853
Epoch 2/3
313/313 ━━━━━━━━━━━━━━━━━━━━ 16s 53ms/step - accuracy: 0.8234 - loss: 0.4081 - val_accuracy: 0.8246 - val_loss: 0.4025
Epoch 3/3
313/313 ━━━━━━━━━━━━━━━━━━━━ 18s 58ms/step - accuracy: 0.9056 - loss: 0.2446 - val_accuracy: 0.8498 - val_loss: 0.4141


#**<font color='red'>4.</font> Evaluate the RNN Model**

In [ ]:
test_loss, test_acc = rnn.evaluate(X_test_padded, y_test)
print("Test accuracy:", test_acc)

782/782 ━━━━━━━━━━━━━━━━━━━━ 10s 12ms/step - accuracy: 0.8389 - loss: 0.4378
Test accuracy: 0.838919997215271


## **Observation**


*   Performance is decent but not optimal with final train accuracy of **90.56%**, Validation accuracy of **85%** and Test accuracy of **83.89%**
*   Increasing sequence length increases training time
*   Model shows some signs of overfitting
*   SimpleRNN captures sequential patterns but struggles with long-term dependencies

